sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamBuiltinAggregations.java

```diff
@@ -39,374 +39,306 @@
+   * {@link CombineFn} for MAX based on {@link Max} and {@link Combine.BinaryCombineFn}.
    */
-  public static final class Count<T> extends BeamSqlUdaf<T, Long, Long> {
-    public Count() {}
+  public static CombineFn createMax(SqlTypeName fieldType) {
+    switch (fieldType) {
+    case INTEGER:
+      return Max.ofIntegers();
+    case SMALLINT:
+      return new ShortMax();
+    case TINYINT:
+      return new ByteMax();
+    case BIGINT:
+      return Max.ofLongs();
+    case FLOAT:
+      return new FloatMax();
+    case DOUBLE:
+      return Max.ofDoubles();
+    case TIMESTAMP:
+      return new DateMax();
+    case DECIMAL:
+      return new BigDecimalMax();
+    default:
+      throw new UnsupportedOperationException(
+          String.format("[%s] is not support in MAX", fieldType));
+  }
+  }
 
-    @Override
-    public Long init() {
-      return 0L;
+  /**
+   * {@link CombineFn} for MAX based on {@link Min} and {@link Combine.BinaryCombineFn}.
+   */
+  public static CombineFn createMin(SqlTypeName fieldType) {
+    switch (fieldType) {
+    case INTEGER:
+      return Min.ofIntegers();
+    case SMALLINT:
+      return new ShortMin();
+    case TINYINT:
+      return new ByteMin();
+    case BIGINT:
+      return Min.ofLongs();
+    case FLOAT:
+      return new FloatMin();
+    case DOUBLE:
+      return Min.ofDoubles();
+    case TIMESTAMP:
+      return new DateMin();
+    case DECIMAL:
+      return new BigDecimalMin();
+    default:
+      throw new UnsupportedOperationException(
+          String.format("[%s] is not support in MIN", fieldType));
+  }
+  }
+
+  /**
+   * {@link CombineFn} for MAX based on {@link Sum} and {@link Combine.BinaryCombineFn}.
+   */
+  public static CombineFn createSum(SqlTypeName fieldType) {
+    switch (fieldType) {
+    case INTEGER:
+      return Sum.ofIntegers();
+    case SMALLINT:
+      return new ShortSum();
+    case TINYINT:
+      return new ByteSum();
+    case BIGINT:
+      return Sum.ofLongs();
+    case FLOAT:
+      return new FloatSum();
+    case DOUBLE:
+      return Sum.ofDoubles();
+    case DECIMAL:
+      return new BigDecimalSum();
+    default:
+      throw new UnsupportedOperationException(
+          String.format("[%s] is not support in SUM", fieldType));
+  }
+  }
+
+  /**
+   * {@link CombineFn} for AVG.
+   */
+  public static CombineFn createAvg(SqlTypeName fieldType) {
+    switch (fieldType) {
+    case INTEGER:
+      return new IntegerAvg();
+    case SMALLINT:
+      return new ShortAvg();
+    case TINYINT:
+      return new ByteAvg();
+    case BIGINT:
+      return new LongAvg();
+    case FLOAT:
+      return new FloatAvg();
+    case DOUBLE:
+      return new DoubleAvg();
+    case DECIMAL:
+      return new BigDecimalAvg();
+    default:
+      throw new UnsupportedOperationException(
+          String.format("[%s] is not support in AVG", fieldType));
+  }
+  }
+
+  static class ShortMax extends Combine.BinaryCombineFn<Short> {
```

Comment:
This is much easier to read now, thank you. However, I think it could still be improved by using common generic implementations for the types that have no existing common implementation, e.g. for Max:

```
  static class UdafMax<T extends Comparable<T>> extends Combine.BinaryCombineFn<T> {
      public T apply(T left, T right) {
          return (right == null || right.compareTo(left) < 0) ? left : right;
      }
  }
```

This could then be used for Short, Byte, Float, Date, and BigDecimal versions, with the standard library CombineFns used for Integer, Long, and Double.

Same goes for all the other aggregators that previously had a single common generic implementation.